src/Framework/AspNetCoreAnalyzers/src/Analyzers/Mvc/DetectAmbiguousActionRoutes.cs

```diff
@@ -0,0 +1,194 @@
+{
+    private static void DetectAmbiguousActionRoutes(SymbolAnalysisContext context, WellKnownTypes wellKnownTypes, List<ActionRoute> actionRoutes)
+    {
+        // Ambiguous action route detection is conservative in what it detects to avoid false positives.
+        //
+        // Successfully matched action routes must:
+        // 1. Be in the same controller.
+        // 2. Have an equivalent route.
+        // 3. Have a matching HTTP method.
+        // 4. Route either be the on the same action or the actions only have known safe attributes that don't impact matching.
+        if (actionRoutes.Count > 0)
+        {
+            // Group action routes together. When multiple match in a group, then report action routes to diagnostics.
+            var groupedByParent = actionRoutes
+                .GroupBy(ar => new ActionRouteGroupKey(ar.ActionSymbol, ar.RouteUsageModel.RoutePattern, ar.HttpMethods, wellKnownTypes));
+
+            foreach (var ambigiousGroup in groupedByParent.Where(g => g.Count() >= 2))
+            {
+                foreach (var ambigiousActionRoute in ambigiousGroup)
+                {
+                    context.ReportDiagnostic(Diagnostic.Create(
+                        DiagnosticDescriptors.AmbiguousActionRoute,
+                        ambigiousActionRoute.RouteUsageModel.UsageContext.RouteToken.GetLocation(),
+                        ambigiousActionRoute.RouteUsageModel.RoutePattern.Root.ToString()));
+                }
+            }
+        }
+    }
+
+    private readonly struct ActionRouteGroupKey : IEquatable<ActionRouteGroupKey>
+    {
+        public IMethodSymbol ActionSymbol { get; }
+        public RoutePatternTree RoutePattern { get; }
+        public ImmutableArray<string> HttpMethods { get; }
+        private readonly WellKnownTypes _wellKnownTypes;
+
+        public ActionRouteGroupKey(IMethodSymbol actionSymbol, RoutePatternTree routePattern, ImmutableArray<string> httpMethods, WellKnownTypes wellKnownTypes)
+        {
+            Debug.Assert(!httpMethods.IsDefault);
+
+            ActionSymbol = actionSymbol;
+            RoutePattern = routePattern;
+            HttpMethods = httpMethods;
+            _wellKnownTypes = wellKnownTypes;
+        }
+
+        public override bool Equals(object obj)
+        {
+            if (obj is ActionRouteGroupKey key)
+            {
+                return Equals(key);
+            }
+            return false;
+        }
+
+        public bool Equals(ActionRouteGroupKey other)
+        {
+            return
+                AmbiguousRoutePatternComparer.Instance.Equals(RoutePattern, other.RoutePattern) &&
+                HasMatchingHttpMethods(HttpMethods, other.HttpMethods) &&
+                CanMatchActions(_wellKnownTypes, ActionSymbol, other.ActionSymbol);
+        }
+
+        private static bool CanMatchActions(WellKnownTypes wellKnownTypes, IMethodSymbol actionSymbol1, IMethodSymbol actionSymbol2)
+        {
+            // Only match routes 
```

Comment:
I was confused by this comment because I was not sure if it is supposed to be a comment related to the `if` statement that comes after it, if so, I suggest something explain that we are checking if the route patterns are described in the same action.